### ISSUE TITLE
MaxDeliveryCount set to int.MaxValue can result in infinite retries

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.Transport.AzureServiceBus#138]
**Describe the bug**

In [ASBL v8 it was decided to set MaxDeliveryCount is to int.MaxValue to ensure messages are not dead-lettered accidentally and to remove the dependency on the endpoints' immediate retries configuration](https://docs.particular.net/transports/upgrades/asb-7to8#queue-and-subscription-maxdeliverycount). This default was copied to ASBS from inception.

- https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/blob/fd490ac2b40b10a8f55d4ba3acbe455706759a81/src/Transport/Administration/QueueCreator.cs#L71
- https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/blob/fd490ac2b40b10a8f55d4ba3acbe455706759a81/src/CommandLine/Queue.cs#L16
- https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/blob/fd490ac2b40b10a8f55d4ba3acbe455706759a81/src/CommandLine/Subscription.cs#L21

However, this can cause issues where messages processing keeps failing. Basically resulting in "infinite" immediate retries. This results in excessive logging and excessive resource utilization.

Workarounds:

Unfortunately, QueueDescription and SubscriptionDescription can no longer be modified with ASBS as was possible with ASBL transport. Queues and subscriptions are created with a fixed MaxDeliveryCount of 2147483647 . Unfortunately, the [ASBS scripting options have the same limitation](https://docs.particular.net/transports/azure-service-bus/operational-scripting).

**It is only possible to override this by manually setting the value for each queue and subscription using Service Bus Explorer: https://github.com/paolosalvatori/ServiceBusExplorer.**

In the users scenarios a `MessageSizeExceededException` occurs, this results in an infinite amount of immediate retries due to `int.MaxValue`. Clearly this is not what you would want. A more sane limit for such scenarios should be used like 10 or 15 which already is a *very* high number for immediate retries. More attempts should happen via delayed retries.

**To Reproduce**
Steps to reproduce the behavior. If you have a repro repository, provide a link.

**Expected behavior**

**Versions:**
 - NuGet package: [e.g. 1.0.0-alpha0073]
 - OS: [e.g. Windows 10 build 1803]
 - .NET Version [e.g. .NET Core 2.1.300]

**Additional context**
- Code repro
- Screenshot

**Comment by SeanFeldman** [comment]
> According to docs.particular.net/transports/upgrades/asb-7to8#queue-and-subscription-maxdeliverycount these seems to be a workaround which requires new deployment

That's legacy ASB documentation and the option is not applicable to this (new) transport.

You're correct in the observation that this _can_ happen. It can happen when using the Standard tier and a message is getting close to the 256KB limit. Arguably, if following Microsoft's recommendation to use the Premium tier in production, chances of having this problem are minimized with a single message limit 1MB. But even then, a message reaching that limit would fail and you'd find yourself in a similar situation. Is this a bug? I don't agree with that. Message size is a concern for most cloud-based systems. If Premium is not used, a system should account for the maximum message size and opt into using the data bus feature to send messages that are too large.

When looking at this issue, I'm not convinced this is a transport concern. The transport does not add those headers. NServiceBus Core does. It does it for this transport, for the legacy ASB, ASQ, SQS, and every other transport. If you want this to be solved, it needs to be done on the platform level, where the additional metadata is introduced and added to the original messages. For that reason, I would suggest replacing this issue with a Core issue.

**Comment by SeanFeldman** [comment]
@ramonsmits was my answer sufficient or you still think there's still a need to keep this issue open?

**Comment by ramonsmits** [comment]
Although your clarification is well written the problem still exists.

Your comment basically states: It can happen that core adds headers during error forwarding causing the messages (header) size to be larger than allowed by the transport.

You state that you consider this a core issue and I partially agree as core should gracefully handle such issues like forwarding it to a poison queue.

It will cause a system to stop processing messages and users will be in the dark on how to resolve this.

Even if core would have graceful handling of this condition this default doesn't make sense. This mechanism exists exactly for this reason to not have systems continuously keep processing a  messages because the processor fails.

**Comment by SeanFeldman** [comment]
> It will cause a system to stop processing messages and users will be in the dark on how to resolve this.

Is this issue raised as a result of a support case?

> Even if core would have graceful handling of this condition this default doesn't make sense. This mechanism exists exactly for this reason to not have systems continuously keep processing a messages because the processor fails.

I don't disagree. The original message failing the processing can be moved to the DLQ. But, have the original message almost on the edge on the maximum size and try to dead-letter it. The message is enriched with the dead-lettering reason that will add to the size and that won't work. How is that different from the scenario with the error queue?

**Comment by ramonsmits** [comment]
@SeanFeldman who would add the DLQ reason? I don't know how this behavior works but if the max delivery count is reached the message is forwarded to the DLQ without any issues. If you state that this still would increase the payload because ASB internally adds a reason that is included in the message size then I believe you.

**Comment by SeanFeldman** [comment]
> who would add the DLQ reason?

The broker/SDK. Not transport.

After looking into the details of the specific incident that has lead to this issue to be opened, it's clear that the originally processed message was on the borderline of the maximum size. This is always a dangerous path and is discouraged. Fat messages are an antipattern, and a claim check pattern ([data bus](https://docs.particular.net/nservicebus/messaging/databus/azure-blob-storage)) can be used to address the problem of size. There's no rule on how big the payload can or cannot be, but getting close to the maximum or not knowing that it can happen is somewhat of a solution design smell to me. This doesn't remove the challenge with NServiceBus additional headers added to the message. They add up to the size. 

An important reminder about Azure Service Bus SDK message - headers are _designed_ to take up to 64KB of the total message size (today's spec). So if the payload is at 200KB+, that's a problem right there. NServiceBus is an opinionated framework. For the recoverability/auditing to work, messages should not be fat. Reduce the size of the message below 192KB, and no issues there. And still, there's another consideration - headers total size. If custom headers do not leave any room for the headers added by NServiceBus, you're back to square one. But that's not the transport decision; that's the framework design the transport has to follow. Until it changes, the transport adheres to the design.

And last, but not least. Assuming this has slipped through the development and QA stages while the transport doesn't have the toggle to adjust an input queue `MaxDeliveryCount` (on purpose), it's still possible to adjust the defaults via Service Bus explorer in the Portal, PowerShell scripting or 3rd party tools.

**Comment by danielmarbach** [comment]
We discussed this during the enhancement release selection process. We do believe given that core does not expose the static recoverability settings and recoverability can be controlled at runtime by recoverability policies there is no way to reasonably predict how many times a message might go through a retry. Therefore setting the delivery count to `int.Max` makes sense from a transport perspective. We are not going to change that

One thing that we might be able to smoothing is the edge case when a message grows in size and thus cannot be delivered to the error queue or audit queue. In such scenarios when a message size exceeded exception happens during the recoverability part of the transport the message could be deadlettered. We think though that could be tricky too because we can't really predict all the edge cases and when we natively deadletter it by passes all the other platform safety mechanisms we provided. Furthermore, we would increase the requirement to monitor the deadletter queues. Since it is a very rare edge case around size limitations that can be addressed by having a proper service tier for production we are currently questioning the return on investment.
